app/managers/security_manager/index.ts

```diff
@@ -0,0 +1,477 @@
+import {Alert, type AlertButton, AppState, type AppStateStatus, Platform} from 'react-native';
+
+import {switchToServer} from '@actions/app/server';
+import {logout} from '@actions/remote/session';
+import {Preferences} from '@constants';
+import {DEFAULT_LOCALE, getTranslations, t} from '@i18n';
+import {getServerCredentials} from '@init/credentials';
+import ManagedApp from '@init/managed_app';
+import {toMilliseconds} from '@utils/datetime';
+import {isMainActivity} from '@utils/helpers';
+import {logError} from '@utils/log';
+
+import type {AvailableScreens} from '@typings/screens/navigation';
+
+type SecurityManagerServerConfig = {
+    Biometrics?: boolean;
+    JailbreakProtection?: boolean;
+    PreventScreenCapture?: boolean;
+    authenticated?: boolean;
+    lastAccessed?: number;
+    siteName?: string;
+};
+
+type SecurityManagerServersCollection = Record<string, SecurityManagerServerConfig>;
+
+class SecurityManagerSingleton {
+    activeServer?: string;
+    serverConfig: SecurityManagerServersCollection = {};
+    backgroundSince = 0;
+    previousAppState?: AppStateStatus;
+    initialized = false;
+
+    constructor() {
+        AppState.addEventListener('change', this.onAppStateChange);
+    }
+
+    /**
+     * Initializes the class with existing servers on app launch.
+     * Should be called when the app starts.
+     */
+    async init(servers: Record<string, ClientConfig>, activeServer?: string) {
+        if (this.initialized) {
+            return;
+        }
+
+        this.initialized = true;
+        const added = new Set<string>();
+        for (const [server, config] of Object.entries(servers)) {
+            if (!this.serverConfig[server]) {
+                this.addServer(server, config);
+                added.add(server);
+            }
+        }
+
+        if (activeServer && (!added.has(activeServer) || !this.activeServer)) {
+            this.activeServer = activeServer;
+            this.setScreenCapturePolicy(activeServer);
+            const isJailbroken = await this.isDeviceJailbroken(activeServer);
+            if (!isJailbroken) {
+                this.authenticateWithBiometricsIfNeeded(activeServer);
+            }
+        }
+    }
+
+    /**
+     * Handles app state changes to prompt authentication when resuming from background.
+     */
+    onAppStateChange = async (appState: AppStateStatus) => {
+        if (this.isAuthenticationHandledByEmm()) {
+            return;
+        }
+
+        const isActive = appState === 'active';
+        const isBackground = appState === 'background';
+
+        if (isActive && this.previousAppState === 'background') {
+            if (this.activeServer) {
+                const config = this.getServerConfig(this.activeServer);
+                if (config && config.Biometrics && isMainActivity()) {
+                    const authExpired = this.backgroundSince > 0 && (Date.now() - this.backgroundSince) >= toMilliseconds({minutes: 5});
```

Comment:
Even if the app is in background for >5 mins, I can still see the push notifications, respond to messages via notification quick replies and potentially even accept incoming calls, all without opening Mattermost, right? I don't know what action sends the app into 'active' state so I might be wrong. Worth checking and documenting.

Reply:
well you could most definitely reply to notifications without opening the app and this PR does not cover the case when sharing content through the share extensions...

But we won't be covering those cases, at least not now.. for the notifications though, the user can set the device to not show the notifications content or actions if not authenticated, but this in fact a user setting on their own device and not controlled by us, not sure if we can prompt for any sort of biometric authentication before sending a reply directly from the notification.

Joining a call on the other hand, you do need to open the app.

The active state is set when the app is in the foreground.

Reply:
Got it, far enough. As long as we know and communicate the limitations, its fine. This will anyways go through security team as well.